File: AppMateriel/physique/lecteurrfiddataserviceimpl.cpp
#include "lecteurrfiddataserviceimpl.h"

Physique::LecteurRFIDDataServiceImpl::LecteurRFIDDataServiceImpl()
{
}

LecteurRFID Physique::LecteurRFIDDataServiceImpl::add(LecteurRFID lecteur)
{
    LecteurRFID newLecteur;
    query.prepare("INSERT INTO lecteur (ip, modele, etatOn) VALUES (:ip, :modele, :etatOn)");
    query.bindValue(":ip", lecteur.getIp());
    query.bindValue(":modele", lecteur.getModele());
    query.bindValue(":etatOn", lecteur.getEtatOn());
    if(query.exec()) {
        newLecteur = getById(query.lastInsertId().toLongLong());
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return newLecteur;
}

bool Physique::LecteurRFIDDataServiceImpl::remove(LecteurRFID lecteur)
{
    bool execute;
    query.prepare("DELETE FROM lecteur WHERE id = :id");
    query.bindValue(":id", (qlonglong)lecteur.getId());
    if(!(execute = query.exec())) {
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return execute;
}

bool Physique::LecteurRFIDDataServiceImpl::update(LecteurRFID lecteur)
{
    bool execute;
    query.prepare("UPDATE lecteur SET ip = :ip, modele = :modele, etatOn = :etatOn WHERE id = :id");
    query.bindValue(":id", (qlonglong)lecteur.getId());
    query.bindValue(":ip", lecteur.getIp());
    query.bindValue(":modele", lecteur.getModele());
    query.bindValue(":etatOn", lecteur.getEtatOn());
    if(!(execute = query.exec())) {
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return execute;
}

QList<LecteurRFID> Physique::LecteurRFIDDataServiceImpl::getAll(int debut, int parPage)
{
    QList<LecteurRFID> lecteurs;
    LecteurRFID lecteur;
    query.prepare("SELECT * FROM lecteur ORDER BY ip ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(query.exec()) {
        while(query.next())
        {
            lecteur = LecteurRFID(query.value(1).toString(), query.value(2).toString(), query.value(3).toBool());
            lecteur.setId(query.value(0).toLongLong());
            lecteurs.append(lecteur);
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return lecteurs;
}

int Physique::LecteurRFIDDataServiceImpl::getCount()
{
    int count = 0;
    query.prepare("SELECT COUNT(*) FROM lecteur");
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

LecteurRFID Physique::LecteurRFIDDataServiceImpl::getById(long id)
{
    LecteurRFID lecteur;
    query.prepare("SELECT * FROM lecteur WHERE id = :id");
    query.bindValue(":id", (qlonglong)id);
    if(query.exec()) {
        if(query.next())
        {
            lecteur = LecteurRFID(query.value(1).toString(), query.value(2).toString(), query.value(3).toBool());
            lecteur.setId(query.value(0).toLongLong());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return lecteur;
}

LecteurRFID Physique::LecteurRFIDDataServiceImpl::getByIp(QString ip)
{
    LecteurRFID lecteur;
    query.prepare("SELECT * FROM lecteur WHERE ip = :ip");
    query.bindValue(":ip", ip);
    if(query.exec()) {
        if(query.next())
        {
            lecteur = LecteurRFID(query.value(1).toString(), query.value(2).toString(), query.value(3).toBool());
            lecteur.setId(query.value(0).toLongLong());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return lecteur;
}

QList<QString> Physique::LecteurRFIDDataServiceImpl::getListIp(int debut, int parPage)
{
    QList<QString> listIp;
    query.prepare("SELECT ip FROM lecteur WHERE ip != '' ORDER BY ip ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(query.exec()) {
        while(query.next()){
            listIp.append(query.value(0).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return listIp;
}

int Physique::LecteurRFIDDataServiceImpl::getCountListIp()
{
    int count = 0;
    query.prepare("SELECT COUNT(ip) FROM lecteur WHERE ip != ''");
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

QList<LecteurRFID> Physique::LecteurRFIDDataServiceImpl::getByModele(QString modele, int debut, int parPage)
{
    QList<LecteurRFID> lecteurs;
    LecteurRFID lecteur;
    query.prepare("SELECT * FROM lecteur WHERE modele = :modele ORDER BY ip ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    query.bindValue(":modele", modele);
    if(query.exec()) {
        while(query.next())
        {
            lecteur = LecteurRFID(query.value(1).toString(), query.value(2).toString(), query.value(3).toBool());
            lecteur.setId(query.value(0).toLongLong());
            lecteurs.append(lecteur);
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return lecteurs;
}

int Physique::LecteurRFIDDataServiceImpl::getCountByModele(QString modele)
{
    int count = 0;
    query.prepare("SELECT COUNT(*) FROM lecteur WHERE modele = :modele");
    query.bindValue(":modele", modele);
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

QList<QString> Physique::LecteurRFIDDataServiceImpl::getListModele(int debut, int parPage)
{
    QList<QString> listModele;
    query.prepare("SELECT DISTINCT modele FROM lecteur WHERE modele != '' ORDER BY modele ASC LIMIT :debut,:parPage"); // Distinc permet de ne pas prendre les valeurs en doublon
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    if(query.exec()) {
        while(query.next()){
            listModele.append(query.value(0).toString());
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return listModele;
}

int Physique::LecteurRFIDDataServiceImpl::getCountListModele()
{
    int count = 0;
    query.prepare("SELECT COUNT(DISTINCT modele) FROM lecteur WHERE modele != ''");
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

QList<LecteurRFID> Physique::LecteurRFIDDataServiceImpl::getByEtatOn(bool etatOn, int debut, int parPage)
{
    QList<LecteurRFID> lecteurs;
    LecteurRFID lecteur;
    query.prepare("SELECT * FROM lecteur WHERE etatOn = :etatOn ORDER BY ip ASC LIMIT :debut,:parPage");
    query.bindValue(":debut", debut);
    query.bindValue(":parPage", parPage);
    query.bindValue(":etatOn", etatOn);
    if(query.exec()) {
        while(query.next())
        {
            lecteur = LecteurRFID(query.value(1).toString(), query.value(2).toString(), query.value(3).toBool());
            lecteur.setId(query.value(0).toLongLong());
            lecteurs.append(lecteur);
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return lecteurs;
}

int Physique::LecteurRFIDDataServiceImpl::getCountByEtatOn(bool etatOn)
{
    int count = 0;
    query.prepare("SELECT COUNT(*) FROM lecteur WHERE etatOn = :etatOn");
    query.bindValue(":etatOn", etatOn);
    if(query.exec()) {
        if(query.next()){
            count = query.value(0).toInt();
        }
    }else{
        qWarning("Error: %s\n", query.lastError().text().toLatin1().data());
    }
    return count;
}

File: LocalisationMateriel/src/java/physique/LecteurRFIDDataServiceImpl.java
package physique;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import metier.LecteurRFID;

public class LecteurRFIDDataServiceImpl implements LecteurRFIDDataService{
    
    private final ServiceSQL sql;
    private Connection cnx;
    private PreparedStatement statement;
    private ResultSet result;

    LecteurRFIDDataServiceImpl() {
        sql = new ServiceSQLImpl();
    }
    
    private void close() throws SQLException, Exception{
        if(!this.result.isClosed()){
            this.result.close();
        }
        if(!this.statement.isClosed()){
            this.statement.close();
        }
        if(!this.cnx.isClosed()){
            this.cnx.close();
        }
        this.sql.close();
    }

    private LecteurRFID createLecteurRFID(ResultSet result) throws Exception {
        long id = result.getLong("id");
        String ip = result.getString("ip");
        String modele = result.getString("modele");
        boolean etatOn = result.getBoolean("etatOn");
        LecteurRFID lecteur = new LecteurRFID(ip, modele, etatOn);
        lecteur.setId(id);
        return lecteur;
    }

    @Override
    public LecteurRFID add(LecteurRFID lecteur) throws Exception {
        cnx = this.sql.connect();
        LecteurRFID newLecteur = null;
        String query = "INSERT INTO lecteur (ip, modele, etatOn) VALUES (?, ?, ?)";
        statement = this.cnx.prepareStatement(query, PreparedStatement.RETURN_GENERATED_KEYS);
        statement.setString(1, lecteur.getIp());
        statement.setString(2, lecteur.getModele());
        statement.setBoolean(3, lecteur.getEtatOn());
        if (statement.executeUpdate() > 0) {
            result = statement.getGeneratedKeys();
            if (result.next()) {
                newLecteur = getById(result.getLong(1));
            }
        }
        this.close();
        return newLecteur;
    }

    @Override
    public boolean remove(LecteurRFID lecteur) throws Exception {
        cnx = this.sql.connect();
        boolean execute = false;
        String query = "DELETE FROM lecteur WHERE id = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setLong(1, lecteur.getId());
        if (statement.executeUpdate() > 0) {
            execute = true;
        }
        this.close();
        return execute;
    }

    @Override
    public boolean update(LecteurRFID lecteur) throws Exception {
        cnx = this.sql.connect();
        boolean execute = false;
        String query = "UPDATE lecteur SET ip = ?, modele = ?, etatOn = ? WHERE id = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, lecteur.getIp());
        statement.setString(2, lecteur.getModele());
        statement.setBoolean(3, lecteur.getEtatOn());
        statement.setLong(4, lecteur.getId());
        if (statement.executeUpdate() > 0) {
            execute = true;
        }
        this.close();
        return execute;
    }

    @Override
    public List<LecteurRFID> getAll(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<LecteurRFID> lecteurs = new ArrayList<LecteurRFID>();
        String query = "SELECT * FROM lecteur ORDER BY ip ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            lecteurs.add(createLecteurRFID(result));
        }
        this.close();
        return lecteurs;
    }

    @Override
    public int getCount() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(*) FROM lecteur";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public LecteurRFID getById(long id) throws Exception {
        cnx = this.sql.connect();
        LecteurRFID lecteur = null;
        String query = "SELECT * FROM lecteur WHERE id = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setLong(1, id);
        result = statement.executeQuery();
        if (result.next()) {
            lecteur = createLecteurRFID(result);
        }
        this.close();
        return lecteur;
    }

    @Override
    public LecteurRFID getByIp(String ip) throws Exception {
        cnx = this.sql.connect();
        LecteurRFID lecteur = null;
        String query = "SELECT * FROM lecteur WHERE ip = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, ip);
        result = statement.executeQuery();
        if (result.next()) {
            lecteur = createLecteurRFID(result);
        }
        this.close();
        return lecteur;
    }

    @Override
    public List<String> getListIp(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<String> listIp = new ArrayList<String>();
        String query = "SELECT ip FROM lecteur WHERE ip != '' ORDER BY ip ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            listIp.add(result.getString("ip"));
        }
        this.close();
        return listIp;
    }

    @Override
    public int getCountListIp() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(ip) FROM lecteur WHERE ip != ''";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public List<LecteurRFID> getByModele(String modele, int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<LecteurRFID> lecteurs = new ArrayList<LecteurRFID>();
        String query = "SELECT * FROM lecteur WHERE modele = ? ORDER BY ip ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, modele);
        statement.setInt(2, debut);
        statement.setInt(3, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            lecteurs.add(createLecteurRFID(result));
        }
        this.close();
        return lecteurs;
    }

    @Override
    public int getCountByModele(String modele) throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(*) FROM lecteur WHERE nomMateriel = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setString(1, modele);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public List<String> getListModele(int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<String> listModele = new ArrayList<String>();
        String query = "SELECT DISTINCT modele FROM lecteur WHERE modele != '' ORDER BY modele ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setInt(1, debut);
        statement.setInt(2, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            listModele.add(result.getString("modele"));
        }
        this.close();
        return listModele;
    }

    @Override
    public int getCountListModele() throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(DISTINCT modele) FROM lecteur WHERE modele != ''";
        statement = this.cnx.prepareStatement(query);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }

    @Override
    public List<LecteurRFID> getByEtatOn(boolean etatOn, int debut, int parPage) throws Exception {
        cnx = this.sql.connect();
        List<LecteurRFID> lecteurs = new ArrayList<LecteurRFID>();
        String query = "SELECT * FROM lecteur WHERE etatOn = ? ORDER BY ip ASC LIMIT ?,?";
        statement = this.cnx.prepareStatement(query);
        statement.setBoolean(1, etatOn);
        statement.setInt(2, debut);
        statement.setInt(3, parPage);
        result = statement.executeQuery();
        while (result.next()) {
            lecteurs.add(createLecteurRFID(result));
        }
        this.close();
        return lecteurs;
    }

    @Override
    public int getCountByEtatOn(boolean etatOn) throws Exception {
        cnx = this.sql.connect();
        int count = 0;
        String query = "SELECT COUNT(*) FROM lecteur WHERE etatOn = ?";
        statement = this.cnx.prepareStatement(query);
        statement.setBoolean(1, etatOn);
        result = statement.executeQuery();
        if (result.next()) {
            count = result.getInt(1);
        }
        this.close();
        return count;
    }
    
}
